### PR TITLE
IGNITE-21868 Moved the sql RO inflights handling from SqlQueryProcess…

### PR DESCRIPTION
**Pull request by denis-chudov**
https://issues.apache.org/jira/browse/IGNITE-21868

Moved the adding/removing the transaction inflights for RO transactions from SqlQueryProcessor to QueryTransactionContext.

### CODE REVIEW

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/QueryTransactionWrapperSelfTest.java
Comment (korlov42) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/QueryTransactionWrapperSelfTest.java:1]: let's add tests to ensure TransactionWrapper/Context interacts with TransactionInflights as expected

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/tx/QueryTransactionContext.java [hunk 1]
```diff
@@ -45,19 +59,26 @@ public QueryTransactionContext(IgniteTransactions transactions, @Nullable Intern
...
+            implicit = true;
+            transaction = (InternalTransaction) transactions.begin(
+                    new TransactionOptions().readOnly(queryType != SqlQueryType.DML));
+        } else {
+            implicit = false;
+            validateStatement(queryType, transaction.isReadOnly());
         }
 
-        validateStatement(queryType, outerTx.isReadOnly());
+        // Adding inflights only for read-only transactions.
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/tx/QueryTransactionContext.java:74]: extended the comment

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/tx/ScriptTransactionContext.java [hunk 2]
```diff
@@ -96,7 +103,11 @@ public CompletableFuture<Void> handleControlStatement(SqlNode node) {
             boolean readOnly = ((IgniteSqlStartTransaction) node).getMode() == IgniteSqlStartTransactionMode.READ_ONLY;
             InternalTransaction tx = (InternalTransaction) queryTxCtx.transactions().begin(new TransactionOptions().readOnly(readOnly));
 
-            this.wrapper = new ScriptTransactionWrapperImpl(tx);
+            if (readOnly && !transactionInflights.addInflight(tx.id(), readOnly)) {
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/tx/ScriptTransactionContext.java:106]: done

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/QueryTransactionWrapperSelfTest.java
Comment (denis-chudov) [reply modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/QueryTransactionWrapperSelfTest.java:1]: added tests

**Review by korlov42 — APPROVED** [review]

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/tx/QueryTransactionContextImpl.java [hunk 3]
```diff
@@ -50,13 +58,23 @@ public QueryTransactionContextImpl(
...
+        QueryTransactionWrapper result;
+
         if (tx == null) {
-            return new QueryTransactionWrapperImpl(
-                    txManager.begin(observableTimeTracker, readOnly), true
-            );
+            transaction = txManager.begin(observableTimeTracker, readOnly);
+            result = new QueryTransactionWrapperImpl(transaction, true, transactionInflights);
+        } else {
+            transaction = tx.unwrap();
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/tx/QueryTransactionContextImpl.java:68]: It is the part of `QueryTransactionWrapperSelfTest#testQueryTransactionWrapperTxInflightsInteraction` regarding `explicitRwTxCtx`. Actually, the only purpose of both these branches is to assign `transaction` and `result` to not duplicate the code below adding the inflight.

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/tx/ScriptTransactionWrapperImpl.java [hunk 4]
```diff
@@ -184,6 +191,10 @@ private void completeTx() {
             default:
                 throw new IllegalStateException("Unknown transaction target state: " + txState);
         }
+
+        if (managedTx.isReadOnly() && completedTx.compareAndSet(false, true)) {
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/tx/ScriptTransactionWrapperImpl.java:195]: It's not, we have an assertion that there are any inflights to remove. Also check the comment above

File: modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/tx/QueryTransactionWrapperImpl.java [hunk 5]
```diff
@@ -42,6 +56,10 @@ public InternalTransaction unwrap() {
 
     @Override
     public CompletableFuture<Void> commitImplicit() {
+        if (transaction.isReadOnly() && committedImplicit.compareAndSet(false, true)) {
```
[the remark this answers is not in the document]
Comment (denis-chudov) [reply modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/tx/QueryTransactionWrapperImpl.java:59]: `getOrStartImplicit` and `commitImplicit` are called inside of SQL code every time when cursor is opened/closed respectively. 
`getOrStartImplicit` may use an existing txn or start a new one.
`commitImplicit` is always used only to commit the explicit transaction on cursor close. It is no-op for explicit txns, but anyway it's called every time when cursor is closed. Unfortunately `commitImplicit` is idempotent and SQL code relies on this fact, `removeInflight` is not - it throws assertion if there are no inflights. So we will have assertion errors in SQL code without this CAS.

**Pull request merged**